File: lista3/cpp/polynomial.cpp
// Marek Traczyński (261748)
// Języki i Paradygmaty Programowania
// Lista 3 Zadanie 2


#include <algorithm>
#include <chrono>
#include <random>
#include "polynomial.h"


/////////////////////////////
// Miscellaneous functions //
/////////////////////////////

template <typename T>
void Polynomial<T>::clearCoeffsZeros()
{
    for (int i = mCoefficients.size() - 1; i > mDegree - 1; i--)
    {
        if (i > 0 && mCoefficients[i] == T(0))
        {
            mCoefficients.pop_back();
            mDegree--;
        }
    }
}

template <typename T>
bool Polynomial<T>::isZeroPoly() const
{
    if (mCoefficients.size() == 1 && mCoefficients[0] == T(0))
    {
        return true;
    }

    return false;
}

template <typename T>
std::tuple<Polynomial<T>, Polynomial<T>> Polynomial<T>::divide(Polynomial<T> divisor) const
{
    if (divisor.isZeroPoly() == true)
    {
        throw "\n\n>> ERROR: Can't divide by 0 polynomial! <<";
    }

    Polynomial<T> result;
    Polynomial<T> remainder = *this;

    while (remainder.isZeroPoly() == false && remainder.mDegree >= divisor.mDegree)
    {
        T coefficient = remainder.mCoefficients.back() / divisor.mCoefficients.back();
        int degree = remainder.mDegree - divisor.mDegree;

        std::vector<T> temp_coeffs(degree + 1);
        temp_coeffs[degree] = coefficient;

        Polynomial<T> temp_poly(temp_coeffs);
        if (temp_poly.isZeroPoly() == true)
        {
            throw "\n\n>> ERROR: Can't divide with the remainder in this type! <<";
        }

        result = result + temp_poly;
        remainder = remainder - (divisor * temp_poly);
    }


    return std::make_tuple(result, remainder);
}


//////////////////
// Constructors //
//////////////////

template <typename T>
Polynomial<T>::Polynomial() : mCoefficients({T(0)}), 
                              mDegree(0) {}

template <typename T>
Polynomial<T>::Polynomial(T value) : mCoefficients({T(value)}), 
                                     mDegree(0) {}

template <typename T>
Polynomial<T>::Polynomial(vector<T> coeffs) : mCoefficients(coeffs),
                                              mDegree(coeffs.size() - 1)
{
    if (coeffs.empty() == true)
    {
        mCoefficients = {T(0)};
        mDegree = 0;
    }

    clearCoeffsZeros();
}


/////////////
// Getters //
/////////////

template <typename T>
vector<T> Polynomial<T>::getCoefficients() const
{
    return mCoefficients;
}

template <typename T>
int Polynomial<T>::getDegree() const
{
    return mDegree;
}


//////////////////////////
// Arithmetic operators //
//////////////////////////

template <typename T>
Polynomial<T> Polynomial<T>::operator + (const Polynomial<T> poly) const
{
    int p1_size = mCoefficients.size();
    int p2_size = poly.mCoefficients.size();
    int result_size = std::max(p1_size, p2_size);

    vector<T> result_coeffs;
    T sum;


    for (int i = 0; i < result_size; i++)
    {
        sum = 0;

        if (i < p1_size)
        {
            sum = sum + mCoefficients[i];
        }
        if (i < p2_size)
        {
            sum = sum + poly.mCoefficients[i];
        }

        result_coeffs.push_back(sum);
    }


    Polynomial<T> result(result_coeffs);

    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator - (const Polynomial<T> poly) const
{
    vector<T> temp_coeffs = poly.mCoefficients;
    std::for_each(std::begin(temp_coeffs), std::end(temp_coeffs), [](T& c){ c = c * T(-1);});
        
    Polynomial<T> temp_poly(temp_coeffs);
    Polynomial<T> result = *this + temp_poly;

    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator * (const Polynomial<T> poly) const
{
    if (this->isZeroPoly() == true || poly.isZeroPoly() == true)
    {
        Polynomial<T> zero_poly;
        return zero_poly;
    }

    int p1_size = mCoefficients.size();
    int p2_size = poly.mCoefficients.size();
    int result_size = p1_size + p2_size - 1;

    vector<T> result_coeffs(result_size);

    for (int i = 0; i < p1_size; i++)
    {
        for (int j = 0; j < p2_size; j++)
        {
            result_coeffs[i+j] = result_coeffs[i+j] + (mCoefficients[i] * poly.mCoefficients[j]);
        }
    }


    Polynomial<T> result(result_coeffs);

    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator / (const Polynomial<T> poly) const
{
    if (poly.isZeroPoly() == true)
    {
        throw "\n\n>> ERROR: Can't divide by 0 polynomial! <<";
    }
    if (this->isZeroPoly() == true)
    {
        Polynomial<T> zero_poly;
        return zero_poly;
    }
    
    Polynomial<T> result = std::get<0>(this->divide(poly));

    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator % (const Polynomial<T> poly) const
{
    Polynomial<T> result;

    if (poly.isZeroPoly() == true)
    {
        throw "\n\n>> ERROR: Can't divide by 0 polynomial! <<";
    }
    if (this->isZeroPoly() == true)
    {
        Polynomial<T> zero_poly;
        return zero_poly;
    }
    
    try
    {
        result = std::get<1>(this->divide(poly));
    }
    catch (const char* msg)
    {
        std::cerr << msg << "\n";
    }
    

    return result;
}

//////////////////////////////////

template <typename T>
Polynomial<T> Polynomial<T>::operator + (const T value) const
{
    vector<T> result_coeffs = mCoefficients;
    result_coeffs[0] = result_coeffs[0] + value;


    Polynomial<T> result(result_coeffs);
    
    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator - (const T value) const
{
    vector<T> result_coeffs = mCoefficients;
    result_coeffs[0] -= value;


    Polynomial<T> result(result_coeffs);
    
    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator * (const T value) const
{
    Polynomial<T> temp_poly(value);
    Polynomial<T> result = *this * temp_poly;
    
    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator / (const T value) const
{
    Polynomial<T> temp_poly(value);
    Polynomial<T> result = *this / temp_poly;
    
    return result;
}

template <typename T>
Polynomial<T> Polynomial<T>::operator % (const T value) const
{
    Polynomial<T> temp_poly(value);
    Polynomial<T> result = *this % temp_poly;
    
    return result;
}


//////////////////////////
// Assignment operators //
//////////////////////////


template <typename T>
Polynomial<T>& Polynomial<T>::operator += (const Polynomial<T> poly)
{
    *this = *this + poly;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator -= (const Polynomial<T> poly)
{
    *this = *this - poly;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator *= (const Polynomial<T> poly)
{
    *this = *this * poly;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator /= (const Polynomial<T> poly)
{
    *this = *this / poly;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator = (const Polynomial<T> poly)
{
    mCoefficients = poly.mCoefficients;
    mDegree = poly.mDegree;

    return *this;
}

//////////////////////////////////

template <typename T>
Polynomial<T>& Polynomial<T>::operator += (const T value)
{
    *this = *this + value;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator -= (const T value)
{
    *this = *this - value;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator *= (const T value)
{
    *this = *this * value;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator /= (const T value)
{
    *this = *this / value;    

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator = (const T value)
{
    Polynomial<T> temp(value);
    *this = temp;

    return *this;
}

template <typename T>
Polynomial<T>& Polynomial<T>::operator = (const vector<T> coeffs)
{
    Polynomial<T> temp(coeffs);
    *this = temp;

    return *this;
}


///////////////////////
// Logical operators //
///////////////////////

template <typename T>
bool Polynomial<T>::operator == (const Polynomial<T> poly) const
{
    return mCoefficients == poly.mCoefficients;
}

template <typename T>
bool Polynomial<T>::operator != (const Polynomial<T> poly) const
{
    return mCoefficients != poly.mCoefficients;
}

template <typename T>
bool Polynomial<T>::operator >= (const Polynomial<T> poly) const
{
    if (mDegree > poly.mDegree)
    {
        return true;
    }

    if (mDegree < poly.mDegree)
    {
        return false;
    }

    
    for (int i = mDegree; i >= 0; i--)
    {
        if (mCoefficients[i] > poly.mCoefficients[i])
        {
            return true;
        }
        else if (mCoefficients[i] < poly.mCoefficients[i])
        {
            return false;
        }
    }
        
    return true; 
}

template <typename T>
bool Polynomial<T>::operator <= (const Polynomial<T> poly) const
{
    if (mDegree < poly.mDegree)
    {
        return true;
    }

    if (mDegree > poly.mDegree)
    {
        return false;
    }

    
    for (int i = mDegree; i >= 0; i--)
    {
        if (mCoefficients[i] < poly.mCoefficients[i])
        {
            return true;
        }
        else if (mCoefficients[i] > poly.mCoefficients[i])
        {
            return false;
        }
    }
        
    return true;
}

template <typename T>
bool Polynomial<T>::operator > (const Polynomial<T> poly) const
{
    if (mDegree > poly.mDegree)
    {
        return true;
    }

    if (mDegree < poly.mDegree)
    {
        return false;
    }

    
    for (int i = mDegree; i >= 0; i--)
    {
        if (mCoefficients[i] > poly.mCoefficients[i])
        {
            return true;
        }
        else if (mCoefficients[i] < poly.mCoefficients[i])
        {
            return false;
        }
    }
        
    return false;    
}

template <typename T>
bool Polynomial<T>::operator < (const Polynomial<T> poly) const
{
    if (mDegree < poly.mDegree)
    {
        return true;
    }

    if (mDegree > poly.mDegree)
    {
        return false;
    }

    
    for (int i = mDegree; i >= 0; i--)
    {
        if (mCoefficients[i] < poly.mCoefficients[i])
        {
            return true;
        }
        else if (mCoefficients[i] > poly.mCoefficients[i])
        {
            return false;
        }
    }
        
    return false;
}

//////////////////////////////////

template <typename T>
bool Polynomial<T>::operator == (const T value) const
{
    if (mCoefficients.size() == 1)
    {
        return mCoefficients[0] == value;
    }

    return false;
}

template <typename T>
bool Polynomial<T>::operator != (const T value) const
{
    if (mCoefficients.size() == 1)
    {
        return mCoefficients[0] != value;
    }

    return false;
}

template <typename T>
bool Polynomial<T>::operator >= (const T value) const
{
    if (mCoefficients.size() == 1)
    {
        return mCoefficients[0] >= value;
    }

    return false;
}

template <typename T>
bool Polynomial<T>::operator <= (const T value) const
{
    if (mCoefficients.size() == 1)
    {
        return mCoefficients[0] <= value;
    }

    return false;
}

template <typename T>
bool Polynomial<T>::operator > (const T value) const
{
    if (mCoefficients.size() == 1)
    {
        return mCoefficients[0] > value;
    }

    return false;
}

template <typename T>
bool Polynomial<T>::operator < (const T value) const
{
    if (mCoefficients.size() == 1)
    {
        return mCoefficients[0] < value;
    }

    return false;
}


//////////////////////////
// Evaluation operators //
//////////////////////////

template <typename T>
T Polynomial<T>::operator () (const T value) const
{
    T result = mCoefficients.back();
    
    for (int i = mDegree - 1; i >= 0; i--)
    {
        result = result * value + mCoefficients[i];
    }

    return result;
}

template <typename T>
T Polynomial<T>::operator [] (const int index) const
{
    if (index > mDegree || index < 0)
    {
        throw "\n\n>> ERROR: Index out of coefficients bounds! <<";
    }

    return mCoefficients[index];
}

File: lista3/java/Polynomial.java
//  Marek Traczyński (261748)
// Języki i Paradygmaty Programowania
// Lista 3 Zadanie 3


import java.util.ArrayList;


public class Polynomial<T extends CoeffType<T>> {

    /////////////
    // Fields  //
    /////////////

    private ArrayList<T> coefficients;
    private int degree;


    //////////////////
    // Constructors //
    //////////////////

    Polynomial(final T value) {
        coefficients = new ArrayList<T>();
        coefficients.add(value);
        degree = 0;
    }
    
    Polynomial(final ArrayList<T> coeffs) {
        coefficients = coeffs;
        degree = coeffs.size() - 1;

        this.clearCoeffsZeros();
    }


    /////////////////////////////
    // Miscellaneous functions //
    /////////////////////////////

    private void clearCoeffsZeros()
    {
        for (int i = coefficients.size() - 1; i > degree - 1; i--)
        {
            if (i > 0 && coefficients.get(i).tIsZero())
            {
                coefficients.remove(i);
                degree--;
            }
        }
    }

    public Tuple<Polynomial<T>, Polynomial<T>> divide(final Polynomial<T> divisor) {
        Polynomial<T> divResult = new Polynomial<T>(this.coefficients.get(0).tZero());
        Polynomial<T> remResult = new Polynomial<T>(this.coefficients);

        while (!(remResult.coefficients.size() == 1 && remResult.coefficients.get(0).tIsZero()) &&
               remResult.degree >= divisor.degree) {
            
            T coefficient = remResult.coefficients.get(remResult.degree).tDiv(divisor.coefficients.get(divisor.degree));
            Polynomial<T> temp = new Polynomial<T>(coefficient);

            for (int i = 0; i < remResult.degree - divisor.degree; i++) {
                temp.coefficients.add(0, temp.coefficients.get(0).tZero());
            }

            divResult = divResult.pAdd(temp);
            remResult = remResult.pSub(divisor.pMul(temp));
        }

        Tuple<Polynomial<T>, Polynomial<T>> resultTuple = new Tuple<Polynomial<T>, Polynomial<T>>(divResult, remResult);

        return resultTuple;
    }


    /////////////
    // Getters //
    /////////////

    public ArrayList<T> getCoefficients() {
        return this.coefficients;
    }

    public int getDegree() {
        return this.degree;
    }
    
    //////////////////////////////////
    // Arithmetic operators methods //
    //////////////////////////////////

    public Polynomial<T> pAdd(final Polynomial<T> poly) {
        int p1Size = this.coefficients.size();
        int p2Size = poly.coefficients.size();
        int resultSize = Math.max(p1Size, p2Size);
        int index = 0;

        T temp;
        ArrayList<T> resultCoeffs = new ArrayList<T>();

        for (int i = 0; i < resultSize; i++) {
            temp = this.coefficients.get(i).tAdd(poly.coefficients.get(i));
            resultCoeffs.add(temp);
            index++;
        }

        while (index < p1Size) {
            resultCoeffs.add(this.coefficients.get(index));
        }
        while (index < p2Size) {
            resultCoeffs.add(poly.coefficients.get(index));
        }

        Polynomial<T> result = new Polynomial<T>(resultCoeffs);

        return result;
    }

    public Polynomial<T> pSub(final Polynomial<T> poly) {
        int p1Size = this.coefficients.size();
        int p2Size = poly.coefficients.size();
        int resultSize = Math.max(p1Size, p2Size);
        int index = 0;

        T temp;
        ArrayList<T> resultCoeffs = new ArrayList<T>();

        for (int i = 0; i < resultSize; i++) {
            temp = this.coefficients.get(i).tSub(poly.coefficients.get(i));
            resultCoeffs.add(temp);
            index++;
        }

        while (index < p1Size) {
            resultCoeffs.add(this.coefficients.get(index));
        }
        while (index < p2Size) {
            resultCoeffs.add(poly.coefficients.get(index));
        }

        Polynomial<T> result = new Polynomial<T>(resultCoeffs);

        return result;
    }

    public Polynomial<T> pMul(final Polynomial<T> poly) {
        int p1Size = this.coefficients.size();
        int p2Size = poly.coefficients.size();
        int resultSize = p1Size + p2Size - 1;
        
        T temp;
        ArrayList<T> resultCoeffs = new ArrayList<T>();

        for (int i = 0; i < resultSize; i++) {
            resultCoeffs.add(this.coefficients.get(i).tZero());
        }
        
        for (int i = 0; i < p1Size; i++) {
            for (int j = 0; j < p2Size; j++) {
                temp = resultCoeffs.get(i+j).tAdd(this.coefficients.get(i).tMul(poly.coefficients.get(j)));
                resultCoeffs.set(i+j, temp);
            }
        }


        Polynomial<T> result = new Polynomial<T>(resultCoeffs);

        return result;
    }

    public Polynomial<T> pDiv(final Polynomial<T> poly) {
        return this.divide(poly).getFirst();
    }

    public Polynomial<T> pMod(final Polynomial<T> poly) {
        return this.divide(poly).getSecond();
    }

    ///////////////////////////////
    // Logical operators methods //
    ///////////////////////////////

    public Boolean pEqual(final Polynomial<T> poly) {
        if (this.degree != poly.degree) {
            return false;
        }

        for (int i = this.degree; i >= 0; i--) {
            if (this.coefficients.get(i).tNotEqual(poly.coefficients.get(i))) {
                return false;
            }
        }
        
        return true;     
    }

    public Boolean pNotEqual(final Polynomial<T> poly) {
        if (this.degree != poly.degree) {
            return true;
        }

        for (int i = this.degree; i >= 0; i--) {
            if (this.coefficients.get(i).tEqual(poly.coefficients.get(i))) {
                return true;
            }
        }
        
        return false;     
    }

    public Boolean pGreaterEq(final Polynomial<T> poly) {
        if (this.degree > poly.degree) {
            return true;
        }

        if (this.degree < poly.degree) {
            return false;
        }


        for (int i = this.degree; i >= 0; i--) {
            if (this.coefficients.get(i).tGreater(poly.coefficients.get(i))) {
                return true;
            } else if (this.coefficients.get(i).tLess(poly.coefficients.get(i))) {
                return false;
            }
        }
        
        return true;    
    }
    
    public Boolean pLessEq(final Polynomial<T> poly) {
        if (this.degree > poly.degree) {
            return true;
        }

        if (this.degree < poly.degree) {
            return false;
        }


        for (int i = this.degree; i >= 0; i--) {
            if (this.coefficients.get(i).tLess(poly.coefficients.get(i))) {
                return true;
            } else if (this.coefficients.get(i).tGreater(poly.coefficients.get(i))) {
                return false;
            }
        }
        
        return true;    
    }
    
    public Boolean pGreater(final Polynomial<T> poly) {
        if (this.degree > poly.degree) {
            return true;
        }

        if (this.degree < poly.degree) {
            return false;
        }


        for (int i = this.degree; i >= 0; i--) {
            if (this.coefficients.get(i).tGreater(poly.coefficients.get(i))) {
                return true;
            } else if (this.coefficients.get(i).tLess(poly.coefficients.get(i))) {
                return false;
            }
        }
        
        return false;    
    }

    public Boolean pLess(final Polynomial<T> poly) {
        if (this.degree > poly.degree) {
            return true;
        }

        if (this.degree < poly.degree) {
            return false;
        }


        for (int i = this.degree; i >= 0; i--) {
            if (this.coefficients.get(i).tLess(poly.coefficients.get(i))) {
                return true;
            } else if (this.coefficients.get(i).tGreater(poly.coefficients.get(i))) {
                return false;
            }
        }
        
        return false;     
    }
    

    //////////////////////////
    // Evaluation operators //
    //////////////////////////

    public T evaluate(final T value) {
        T result = this.coefficients.get(this.degree);
        T temp;
        
        for (int i = this.degree - 1; i >= 0; i--) {
            temp = result.tMul(value);
            result = temp.tAdd(coefficients.get(i));
        }

        return result;
    }

    public T getCoeff(final int index) {
        return this.coefficients.get(index);
    }


    /////////////////////
    // toString method //
    /////////////////////

    public String toString() {
        String result = "";
        String plus = "";
        String space = "";
        
        if (this.coefficients.size() == 1 && this.coefficients.get(0).tIsZero()) {
            return "0";
        }

        for(int i = this.degree; i >= 0; i--) {
            if (this.coefficients.get(i).tIsZero()) {
                continue;
            }

            if (this.coefficients.get(i).tGreater(this.coefficients.get(i).tZero())) {
                result += space + plus + space;
            } else if (this.coefficients.get(i).tLess(this.coefficients.get(i).tZero())) {
                result += space + "-" + space;
            }
            
            plus = "+";
            space = " ";

            if (i == 0 || (this.coefficients.get(i).tAbsolute()).tNotEqual(this.coefficients.get(i).tOne())) {
                result += this.coefficients.get(i).tAbsolute();
            }
            
            if (i == 0)
            {
                continue;
            }

            result += "x";

            if (i == 1)
            {
                continue;
            }

            result += "^" + i;
        }

        return result;
    }
}
